### PR TITLE
Improve conversion to PURL

### PR DESCRIPTION
**Pull request by ken-duck**
Improve package conversion to PURL

This pull request makes the following changes:
* Update convertGopkgNameToPurl to generate better PURLs

### CODE REVIEW

File: packages/packages.go [hunk 1]
```diff
@@ -28,12 +32,22 @@ type Packages interface {
...
+	case githubPattern.MatchString(name):
+		// Currently OSS Index's github format support is based on repository
+		// owner/name, so restrict the PURL name path.
+		//
+		// Once golang format support is improved we can switch to that and get
+		// more precise and refined results on repository sub-paths
+		name = githubPattern.ReplaceAllString(name, "github/$1/$2")
+
+	case gopkg2Pattern.MatchString(name):
+		name = gopkg2Pattern.ReplaceAllString(name, "github/$1/$2")
```
Comment (DarthHater) [inline packages/packages.go]: QUESTION: Not knowing enough here, but this uses the gopk2Pattern, which is: `("^gopkg.in/([^/]+)/([^.]+).*")`, should it be stuffing `github` in the replace?

File: packages/packages.go [hunk 2]
```diff
@@ -28,12 +32,22 @@ type Packages interface {
...
+
+	case gopkg1Pattern.MatchString(name):
+		name = gopkg1Pattern.ReplaceAllString(name, "github/go-$1/$1")
```
Comment (DarthHater) [inline packages/packages.go]: QUESTION: Not knowing enough here, but this uses the gopkg1Pattern, which is: `"^gopkg.in/([^.]+).*"`, should it be stuffing `github` in the replace?

File: packages/packages.go (hunk 1 above)
Comment (ken-duck) [reply packages/packages.go]: gopkg.in is a service that maps shortened URLs with github repositories following a certain pattern, so we are just following the same pattern to find the real URLs.

File: packages/packages.go (hunk 2 above)
Comment (ken-duck) [reply packages/packages.go]: Same as above.

For some additional flavouring, a "proper" golang PURL for github looks like this: `pkg:golang/github.com/<owner>/<name>@<version>`

However, OSS Index has better support for github PURLs than golang PURLs at the moment, though improvements are in progress. For that reason we use github PURLs instead: `pkg:github/<owner>/<name>@<version>`

**Pull request merged**